### DISCUSSION TITLE
How do the refresh tokens used for auth work?

### DISCUSSION THREAD

**Discussion by He1nr1chK** [supabase/supabase#2204] [Questions] [answered]
Hi @kangmingtay , I am trying to use a refresh token to authenticate a user using the signIn() method. Currently I am using the method suggested by @jpstrikesback of calling AuthSession.startAsync() but this for a number of reasons this is not ideal. I would like to know whether I can directly use the refresh token I get from the OAuth provider to sign in or if the refresh token has to pass through the GoTrue server first (does it get mutated maybe?). When trying to sign in at the moment using the above mentioned method I get the error: Invalid Refresh Token. If it does indeed first need to pass through the GoTrue server, is there an endpoint I can call with the data I receive back from the OAuth provider to get the necessary refresh token back to be used for auth?

**Comment by kangmingtay** [comment, marked as answer]
Hi @He1nr1chK, thanks for your question!

>  I would like to know whether I can directly use the refresh token I get from the OAuth provider to sign in or if the refresh token has to pass through the GoTrue server first (does it get mutated maybe?). 

Yeah you can't use the refresh token provided by the OAuth provider to sign in. To sign in with an OAuth provider, it is recommended to use `supabase.auth.signIn({ provider }}`. On a successful sign in, the gotrue server will return an access token, refresh token and a provider token. The provider token corresponds with the token you receive from the OAuth provider to request for the user's data while the access token and refresh token are created on gotrue's end. 

> When trying to sign in at the moment using the above mentioned method I get the error: Invalid Refresh Token. If it does indeed first need to pass through the GoTrue server, is there an endpoint I can call with the data I receive back from the OAuth provider to get the necessary refresh token back to be used for auth? 

It seems like you have your own way of handling the sign in with an OAuth provider and you want to pass the data to the `auth.users` table. Currently, there isn't an endpoint you can call with the data received back from the OAuth provider to get the gotrue refresh token because gotrue has no way to validate that you have signed in with an OAuth provider successfully. 

Hope this clarifies some things! Feel free to let me know if you have any more questions.

**Reply by He1nr1chK** [reply]
Thanks for the swift reply. Yes I would handle the sign in and then I get the accessToken, refreshToken etc back from the OAuth provider. I assume its these values that the GoTrue server uses internally to generate the refresh_token that it returns from the authorize/callback methods or am I mistaken?

**Reply by kangmingtay** [reply]
Actually, the gotrue server generates its own refresh token that is independent of the access token and refresh token from the OAuth provider. Hope this helps!

**Reply by roker15** [reply]
I see here in the link that there is no event for jwt expiration.
#3673 (comment)

If there is not JWT token expiration event fired from onAuthstatechanged then how session is maintained and updated when token is automatically refreshed in background?

**Reply by joshgeller** [reply]
When the JWT expires, I see an event called `TOKEN_REFRESHED` in `onAuthStateChanged`, followed immediately by `SIGNED_IN`.
